includes/Modules/Sign_In_With_Google.php

```diff
@@ -82,23 +140,58 @@ public function handle_google_auth() {
 			exit;
 		}
 
-		$client_id = $this->get_settings()->get()['clientID'];
-		$id_token  = $this->context->input()->filter( INPUT_POST, 'credential' );
+		$id_token = $this->context->input()->filter( INPUT_POST, 'credential' );
 		try {
-			$client  = new Google_Client( array( 'client_id' => $client_id ) );
-			$payload = $client->verifyIdToken( $id_token );
+			$payload = $this->client->verifyIdToken( $id_token );
+		} catch ( \Exception $e ) {
+			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+			exit;
+		}
 
-			if ( empty( $payload ) ) {
-				wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
-				exit;
-			}
+		if ( is_null( $payload ) || empty( $payload ) || ! array_key_exists( 'sub', $payload ) || empty( $payload['sub'] ) || ! array_key_exists( 'email', $payload ) || empty( $payload['email'] ) ) {
+			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+			exit;
+		}
 
-			// @TODO implement further flow using $payload in #9339.
+		$google_user_id    = $payload['sub'];
+		$google_user_email = $payload['email'];
 
-		} catch ( \Exception $e ) {
-			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+		// Check if there are any existing WordPress users connected to this Google account.
+		// The user ID is used as the unique identifier because users can change the email on their Google account.
+		$existing_users = get_users(
+			array(
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_key
+				'meta_key'   => self::SIGN_IN_WITH_GOOGLE_USER_ID_OPTION,
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+				'meta_value' => $google_user_id,
+				'number'     => 1,
+			)
+		);
+		if ( ! empty( $existing_users ) && $existing_users[0] instanceof WP_User ) {
+			return $this->login_user( $existing_users[0] );
+		}
+
+		// Find an existing user that matches the email and link to their Google account by store their user ID in user meta.
+		$existing_user = get_user_by( 'email', $google_user_email );
+		if ( $existing_user instanceof WP_User ) {
+			add_user_meta( $existing_user->ID, self::SIGN_IN_WITH_GOOGLE_USER_ID_OPTION, $google_user_id, true );
+			return $this->login_user( $existing_user );
+		}
+
+		// Create a new user if "Anyone can register" setting is enabled.
+		$registration_open = get_option( 'users_can_register' );
+
+		if ( $registration_open ) {
+			$new_user_id = wp_create_user( $google_user_email, wp_generate_password(), $google_user_email );
+			add_user_meta( $new_user_id, self::SIGN_IN_WITH_GOOGLE_USER_ID_OPTION, $google_user_id, true );
+			return $this->login_user( get_user_by( 'id', $new_user_id ) );
+		} else {
+			wp_safe_redirect( add_query_arg( 'error', 'registration_disabled', wp_login_url() ) );
```

Comment:
Instead of using `if | else` let's check for invalid condition and return early first, example
```php
if ( ! $registration_open ) {
    wp_safe_redirect( add_query_arg( 'error', 'registration_disabled', wp_login_url() ) );
    exit;
}
// create user and do login if registration is open
```

Reply:
We should use the `google_auth_failed` code instead of `registration_disabled`.